crates/btc-relay/src/types.rs

```diff
@@ -45,3 +49,85 @@ impl<AccountId, BlockNumber> RichBlockHeader<AccountId, BlockNumber> {
         })
     }
 }
+
+pub struct OpReturnPaymentData<T: frame_system::Config> {
+    pub op_return: H256,
+    // vec of (amount, address)
+    payments: Vec<(i64, BtcAddress)>,
+    _marker: sp_std::marker::PhantomData<T>,
+}
+
+impl<T: crate::Config> OpReturnPaymentData<T> {
+    pub fn try_from_transaction(transaction: Transaction) -> Result<Self, DispatchError> {
+        // check the number of outputs - this check is redundant due to the checks below, but
+        // this serves to put an upperbound to the number of iterations
+        ensure!(
+            transaction.outputs.len() <= ACCEPTED_MAX_OPRETURN_TRANSACTION_OUTPUTS as usize,
+            Error::<T>::MalformedTransaction
+        );
+
+        let mut payments = Vec::new();
+        let mut op_returns = Vec::new();
```

Comment:
Could adopt an enum here (i.e. `PaymentOrData`), then we already know the expected size of the vector.

Reply:
Are you worried about performance? We could initialize payments to size (num_outputs -1) and op_returns to 1. Having an enum would complicate the logic below

Reply:
I think an enum would improve both performance and readability

Reply:
I actually think it'd probably be worse - after creating `PaymentOrData`, you need to iterate over them again to find where the op_return is. Then, removing it from the vec would be `O(n-i)`, and if you don't remove it, you decrease the readability of the rest of code.

Reply:
We could match on the enum vec, but point taken. Just wondering if there's a more streamlined solution here